### PR TITLE
[proofs] [sat] Handle resolution proofs for optimized clauses in incremental

### PR DESCRIPTION
**Pull request by HanielB**
The SAT solver may associate, to the result of a resolution, a lower assertion
level than the current one. To be able to produce proofs to such clauses, we
extend the SAT proof manager to track which resolution results were
optimized. Then, when the SAT solver pops, the manager is notified and stores
the proofs for these optimized clauses, so that they can be reinserted when the
user context is popped. The latter is handled by the SAT proof manager's
`OptClauseManager` attribute.

### CODE REVIEW

File: src/prop/sat_proof_manager.h [hunk 1]
```diff
@@ -582,6 +587,22 @@ class SatProofManager
...
+
+  /** User-context dependent map from resolution conclusions to their assertion
+      level. */
+  context::CDHashMap<Node, int> d_optResLevels;
+  /** User-context-dependent map assertion level to proof nodes.
+   *
+   * This map is used to update the internal proof of this manager when the
+   * context pops.
+   */
+  std::map<int, std::vector<std::shared_ptr<ProofNode>>> d_optResProofs;
```
Comment (ajreynol) [inline src/prop/sat_proof_manager.h:602]: This is not context-dependent?

File: src/prop/sat_proof_manager.h [hunk 2]
```diff
@@ -582,6 +587,22 @@ class SatProofManager
   Node getClauseNode(const Minisat::Clause& clause);
   /** Prints clause, as a sequence of literals, in the "sat-proof" trace. */
   void printClause(const Minisat::Clause& clause);
+
+  /** The user context */
```
Comment (ajreynol) [inline src/prop/sat_proof_manager.h:591]: Is it possible to make this into EnvObj?  Otherwise we will have to clean this later.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/prop/sat_proof_manager.h (hunk 1 above)
Comment (HanielB) [reply src/prop/sat_proof_manager.h:602]: No, the documentation was wrong. This map is used by the optimized clauses manager, which is reinserting stuff in its parent proof from this map, according to how the user levels are popped.

File: src/prop/sat_proof_manager.h (hunk 2 above)
Comment (HanielB) [reply src/prop/sat_proof_manager.h:591]: Will do

**Review by ajreynol — APPROVED** [review]

**Pull request merged**